keeper-gradle-plugin/src/main/java/com/slack/keeper/KeeperPlugin.kt

```diff
@@ -100,95 +102,184 @@ class KeeperPlugin : Plugin<Project> {
     check(gradleVersion >= MIN_GRADLE_VERSION) {
       "Keeper requires Gradle 6.0 or later."
     }
-    with(project) {
-      pluginManager.withPlugin("com.android.application") {
-        val extension = project.extensions.create<KeeperExtension>("keeper")
-
-        // Set up r8 configuration
-        val r8Configuration = configurations.create(CONFIGURATION_NAME) {
-          description = "R8 dependencies for Keeper. This is used solely for the PrintUses CLI"
-          isVisible = false
-          isCanBeConsumed = false
-          isCanBeResolved = true
-          defaultDependencies {
-            add(project.dependencies.create("com.android.tools:r8:$DEFAULT_R8_VERSION"))
-          }
-        }
+    project.pluginManager.withPlugin("com.android.application") {
+      val appExtension = project.extensions.getByType<AppExtension>()
+      val extension = project.extensions.create<KeeperExtension>("keeper")
+      val diagnosticOutputDir = project.layout.buildDirectory.dir("$INTERMEDIATES_DIR/diagnostics")
+      project.configureKeepRulesGeneration(appExtension, extension, diagnosticOutputDir)
+      project.configureL8Rules(appExtension, extension, diagnosticOutputDir)
+    }
+  }
 
-        val appExtension = project.extensions.getByType<AppExtension>()
+  private fun Project.configureL8Rules(
+      appExtension: AppExtension,
+      extension: KeeperExtension,
+      diagnosticOutputDir: Provider<Directory>
+  ) {
+    project.afterEvaluate {
+      if (extension.enableL8RuleSharing.getOrElse(false)) {
+        val r8Enabled = !project.hasProperty("android.enableR8") ||
+            project.property("android.enableR8")?.toString()?.toBoolean() != false
+        if (r8Enabled) {
+          appExtension.onApplicableVariants(project, extension) { testVariant, appVariant ->
+            val inputFiles = tasks.withType<R8Task>()
+                .named("minify${appVariant.name.capitalize(Locale.US)}WithR8")
```

Comment:
Couldn't you also just do 
```
val inputFiles = tasks.named<R8Task>("minify${appVariant.name.capitalize(Locale.US)}WithR8")
```

I worry about usages of `withType()` because it's easy to accidentally realize tasks if you then use the wrong method on the result.

Reply:
TIL!

Reply:
ecaf66f